src/shortest_path/mod.rs

```diff
@@ -1089,3 +1091,117 @@ pub fn graph_distance_matrix(
     );
     matrix.into_pyarray(py).into()
 }
+
+/// Return the average shortest path length for a :class:`~retworkx.PyDiGraph`
+/// with unweighted edges.
+///
+/// The average shortest path length is calculated as
+///
+/// .. math::
+///
+///     a =\sum_{s,t \in V} \frac{d(s, t)}{n(n-1)}
+///
+/// where :math:`V` is the set of nodes in ``graph``, :math:`d(s, t)` is the
+/// shortest path length from :math:`s` to :math:`t`, and :math:`n` is the
+/// number of nodes in ``graph``.
+///
+/// This function is also multithreaded and will run in parallel if the number
+/// of nodes in the graph is above the value of ``parallel_threshold`` (it
+/// defaults to 300). If the function will be running in parallel the env var
+/// ``RAYON_NUM_THREADS`` can be used to adjust how many threads will be used.
+/// By default it will use all available CPUs if the environment variable is
+/// not specified.
+///
+/// :param PyDiGraph graph: The graph to compute the average shortest path length
+///     for
+/// :param int parallel_threshold: The number of nodes to calculate the
+///     the distance matrix in parallel at. It defaults to 300, but this can
+///     be tuned to any number of nodes.
+/// :param bool as_undirected: If set to ``True`` the input directed graph
+///     will be treated as if each edge was bidirectional/undirected while
+///     finding the shortest paths. Default: ``False``.
+///
+/// :returns: The average shortest path length. If the graph is empty this
+///     will return NaN, if there is a single node 0 will be returned and if
+///     the graph is disconnected it will return infinity.
+/// :rtype: float
+#[pyfunction(parallel_threshold = "300", as_undirected = "false")]
+#[pyo3(
+    text_signature = "(graph, /, parallel_threshold=300, as_undirected=False)"
+)]
+pub fn digraph_unweighted_average_shortest_path_length(
+    graph: &digraph::PyDiGraph,
+    parallel_threshold: usize,
+    as_undirected: bool,
+) -> f64 {
+    let n = graph.node_count();
+    if n == 0 {
+        return std::f64::NAN;
+    }
+    if n == 1 {
+        return 0.0;
+    }
+    if !is_connected(&graph.graph) {
```

Comment:
We need to rework the logic here for directed graphs since `is_connected` only checks if the first node is connected to all other nodes of the digraph (which does not really make any sense).

I see several possibilities depending on the definition we actually want for this function:
- Check `is_weakly_connected` if `as_undirected=True` and `is_strongly_connected` if `as_undirected=False`
- Check only `is_weakly_connected` as Networkx does and for disconnected pairs define the distance to be zero. This is not really consistent in my opinion.
- Check nothing and for disconnected pairs define the distance to be zero. This is the definition i found in [wikipedia](https://en.wikipedia.org/wiki/Average_path_length). A possible improvement here would be to divide by the number of connected pairs of nodes instead of `n`.

Reply:
Nice catch! Yeah, this is clearly not correct. I also wasn't happy with the networkx weakly connected approach so I tried to do it differently, but you're right this doesn't make much sense. I kind of like the approach on Wikipedia since it means we only traverse the graph once, I'll give it a go and update it with this approach tomorrow.

Reply:
I've reworked this in https://github.com/Qiskit/retworkx/pull/403/commits/44907620bc88ca3d48227b3cebccae0e4e1fb975 to use the wikipedia definition and added a test comparing a partially disconnected graph with the fully connected equivalent.